### PR TITLE
User-based ObjectStore

### PR DESCRIPTION
**Pull request by VJalili**
This PR is in accordance with the _Federated user-based ObjectStore_ goal. The updates introduced in this PR are mock-up of the _Federated user-based ObjectStore_, and can generally be grouped as: 

1. **User-based ObjectStore**. Current ObjectStore reads its configuration (e.g., provider, bucket name, access and secret keys) from `config/object_store_conf.xml` and applies them Galaxy instance-wide. In other words, the data uploaded/created/processes by any of the users are written-to or read-from a common bucket to which the Galaxy instance has full access (accordingly, implicitly accessible by all the users of the galaxy instance). This PR introduces **user-based objectstore**, where each user defines his/her own `PluggedMedia` (e.g., AWS S3 bucket(s), Azure Blob, and etc.) where a `dataset` is read/written to that `PluggedMedia`.

2. **Provider agnostic ObjectStore**. This PR replaces the `S3ObjecStore` module with `cloud` module. The former uses `boto` and provides access to AWS S3 only; while, the latter leverages on `CloudBridge` which provides an interface for unified access to various providers (including S3 and Swift).

What is expected in future related PRs: 
- at this PR, only `cloud` module is user-based; other modules such as `DistributedObjectStore` are not (user should be able to define how to distribute his/her own data on the `PluggedMedia`s he/she has defined using a user-based `DistributedObjectStore`).

- some UI elements will be defined to allow user (a) define a `PluggedMedia`, (b) choose a `PluggedMedia` when uploading/creating a new dataset. At this PR, to define a `PluggedMedia`, one should manually add it to the database (`PluggedMedia` table). Additionally, the code has two temporary code-snippets (marked with `TEMP BLOCK`) which choose the first available `PluggedMedia` of the user.

### CODE REVIEW

**Comment by bgruening** [comment]
@VJalili regarding user-preferences you might want to look at https://github.com/galaxyproject/galaxy/pull/3383

This is an awesome feature, looking so much forward to use it - Thanks!

File: database/info.txt [hunk 1]
```diff
@@ -1 +0,0 @@
-This folder contains the data
```
Comment (nsoranzo) [inline database/info.txt]: Please revert this change, this file is used to ensure the `database` directory is present when cloning the Git repo.

File: lib/galaxy/jobs/__init__.py [hunk 2]
```diff
@@ -902,10 +902,16 @@ def get_special( ):
     def _create_working_directory( self ):
         job = self.get_job()
         try:
+            # TEMP BLOCK --- START
+            pluggedMedia = None
```
Comment (jmchilton) [inline lib/galaxy/jobs/__init__.py]: Everywhere ``pluggedMedia`` appears as a variable name and/or function input parameter it should be called ``plugged_media``. The Python convention is to use camel case (e.g. ``PluggedMedia`` just for class names.

File: lib/galaxy/jobs/__init__.py [hunk 3]
```diff
@@ -902,10 +902,16 @@ def get_special( ):
...
+            for pM in job.user.pluggedMedia:
```
Comment (jmchilton) [inline lib/galaxy/jobs/__init__.py]: This ``for`` loop just to set a variable is a bit awkward. I think it'd be better to add a method to the ``User`` class in ``lib/galaxy/model/__init__.py`` that defines this.

```
def current_plugged_media(self):
    return None if not len(self.plugged_media) else self.plugged_media[-1]
```

and then access like:

```
job.user.get_current_plugged_media()
```

It is really unfortunate that ``plugged_media`` is also used as both a list and a single instance. I'd consider a name that has a more obvious, distinct singular vs. plural modality (e.g. ``plugged_volume`` vs. ``plugged_volumes``) or use ``plugged_medium`` for the single instance you are fetching.

File: lib/galaxy/objectstore/cloud.py [hunk 4]
```diff
@@ -620,19 +687,20 @@ def get_store_usage_percent(self):
...
+class SwiftObjectStore(CloudObjectStore):
     """
     Object store that stores objects as items in a Swift bucket. A local
     cache exists that is used as an intermediate location for files between
     Galaxy and Swift.
     """
 
     def _configure_connection(self):
+        # TODO: Replace with Cloudbridge connection.
         log.debug("Configuring Swift Connection")
```
Comment (jmchilton) [inline lib/galaxy/objectstore/cloud.py:699]: Can you mark this PR as a WIP until this TODO is addressed?

Comment (jmchilton) [reply lib/galaxy/objectstore/cloud.py:699]: If it is already addressed in the super class - can you just delete the code and the comment?

**Comment by jmchilton** [comment]
This is exciting functionality and a great direction to move in - thanks for the contribution!

That said, I'd really like to see some large structural changes to this PR - in particular I'd like to see this as 3 different PRs so that I can comment on the pieces individually.

The three PRs I'd like to see are:

- First a very small easy-to-merge PR that addresses all the smaller white space changes and such you make. If those are important to you - we can just merge those quickly and it would really reduce the visual clutter for the other PRs. This is a pattern I frequently follow before making big changes - there is an initial PR to smooth out and clarify the pieces of code I'm going to touch. This would make the other two PRs more atomic as well.
- Secondly, I'd like to see the swap to cloudbridge as its own PR - that should be doable without the pluggable media stuff right? I mean there is an open PR from you about that right? I thought it required more work on your end - but if you are waiting on committers maybe we can have a small team meeting to discuss getting that merged - I don't think anyone wants to block your progress. That said - based on TODOs and such - this part of the PR seems like it needs some work - restoring swift functionality for instance.

  One thing to consider here is to keep the existing object stores as they are - and just add a new cloud one. One potential criticism of what you have done here is that you break existing object store confs by changing the names and stuff. If you just added your own object store first and even insisted that it is used for the new functionality I think that would be fine. Everyone could be more confident about merging it then if we knew it wasn't breaking old stuff. Longer term that duplication is problematic - but it could be updated in an atomic PR after this hypothetical one to just provide the backward compatibility endpoints to make swift and s3 point at some specialization of the cloud provider.
- The third PR would be the pluggable media PR. And for that PR I'd like to see it marked as WIP until a full demo is ready to go - even if it is just a set of API calls that the reviewer can imaging a GUI sitting on top of. I don't feel comfortable making these big changes to the object store interaction code and to the database until I definitely see all the pieces fitting together. Maybe everything that is needed to make this usable today is there and I just need it pointed out to me - but if not I'd like to delay if possible and make the PR feel more atomic. I care passionately about the interaction between the job runners and object stores and I think probably others do as well and this adds some complexity there - given that a lot of people care about it - it would be better to make a big change atomically I think so we can be sure this is the right complexity. And separating the two PRs discussed above would give everyone a better shot at commenting intelligently about the work as you go I think.

  This is a lot to ask I think - to keep so many big changes sort of in WIP branches - but I do think it is the best way to go. I started my collections branch with database changes just like this in like December of 2014 and didn't merge the final collections PR until maybe May of 2015 once the API and tool interactions, etc... were just right (or right enough - demonstrable at least). I started a branch in the fall of 2015 for CWL support that included database interactions and I'm still working on producing something atomic that feels correct and complete.

All that said - I'm not a -1 or anything - this isn't something I'm intending to force on you. I'm just letting you know how I would approach it and how I would feel more comfortable reviewing it. Good luck!

**Comment by VJalili** [comment]
@bgruening thanks for the reference. Indeed having a section in user preferences where the user can define PluggedMedia (e.g., AWS S3, Azure Blob, and etc.) is required to improve usability of user-based objectstore. Additionally that section should allow user to define hierarchical relation between his/her various plugged medias (similar to [object_store_config](https://github.com/galaxyproject/galaxy/blob/dev/config/object_store_conf.xml.sample)).

File: database/info.txt (hunk 1 above)
Comment (VJalili) [reply database/info.txt]: Thanks, done.

File: lib/galaxy/jobs/__init__.py (hunk 2 above)
Comment (VJalili) [reply lib/galaxy/jobs/__init__.py]: Thanks, refactored.

File: lib/galaxy/jobs/__init__.py (hunk 3 above)
Comment (VJalili) [reply lib/galaxy/jobs/__init__.py]: As marked, this part of the code is temporary. There will be some UI's asking user which `plugged_media` to use, or allow user to set a default `plugged_media` from preferences. However, till then, we're assuming that the user has only one `plugged_media` and we're retrieving it this way. 

regarding naming; indeed `job.user.plugged_media` is a relation to the `plugged_media` table.

**Comment by jgoecks** [comment]
+1 on @jmchilton's suggestions. These will help in scoping/modularizing/understanding this excellent work and generating digestable PRs.

**Comment by VJalili** [comment]
are you suggesting to close this PR, separate cloudbridge, and create a new PR when use-based ObjectStore is fully functional ?

**Comment by dannon** [comment]
I didn't want to jump in on this PR right away since I feel like we've talked about this in person quite a bit, but I agree pretty much completely with what @jmchilton suggests, which closely echoes what I feel like we've talked about already.

Small logical units of incremental change are *vastly* preferable and are much easier to understand and review/approve, or suggest improvements for.  I like the three-PR breakdown John suggests -- one stylistic, with no substantial changes, for if you need/want to fix style in places prior to work.  Then the cloudbridge part on existing components.  Finally, new PluggableMedia changes.

I'll also echo, as I've mentioned several times in person, that we should *not* just throw away S3ObjectStore without a dead-simple automated upgrade path where existing configurations will continue to work (this is general practice for *anything* changing in Galaxy configuration).  The easiest solution here is probably to, instead of renaming S3ObjectStore, just leave it as-is for now and add the CloudObjectStore separately.

**Comment by jxtx** [comment]
> I'll also echo, as I've mentioned several times in person, that we should not just throw away S3ObjectStore without a dead-simple automated upgrade path where existing configurations will continue to work (this is general practice for anything changing in Galaxy configuration). The easiest solution here is probably to, instead of renaming S3ObjectStore, just leave it as-is for now and add the CloudObjectStore separately.

I also don't understand the reason for removing it. If we just leave it we have backward compatibility without needing to write any conversion scripts or whatever, which is what we usually strive for.

**Comment by VJalili** [comment]
What do you think about the user-based objectstore concept this PR is introducing?

**Comment by VJalili** [comment]
Thanks for your comments. Accordingly, I created PRs [#135](https://github.com/galaxyproject/starforge/pull/135) and [#136](https://github.com/galaxyproject/starforge/pull/136) on starforge, and PRs [#4333](https://github.com/galaxyproject/galaxy/pull/4333), [#4342](https://github.com/galaxyproject/galaxy/pull/4342) and [4352](https://github.com/galaxyproject/galaxy/pull/4352) on galaxy.

**Comment by nsoranzo** [comment]
Thanks @VJalili, I'm closing this one then.

**Comment by VJalili** [comment]
@nsoranzo , The other PRs are focused on `Cloud` ObjectStore, and this PR extends it as a User-based ObjectStore; i.e., adding a user-defined "PluggedMedia" (e.g., AWS S3 bucket, Azure blob). The main difference between the current ObjectStore (e.g., `S3ObjectStore`) and the User-based ObjectStore (which this PR introduces) is as follows: 

- in the current objectstore, a configured media (e.g., AWS S3) is used for the all the galaxy users. In other words, only one S3 bucket per galaxy instance.

- in the User-based objectstore, each user defines his/her own media (e.g., his very own AWS S3 bucket, or Google drive).  In other words, many S3 buckets per galaxy instance. 

The `Cloud` is submitted as a separated PR following @jmchilton @jgoecks and @dannon suggestion; but those PRs are not adding User-based ObjectStore. Having merged the aforementioned PRs, we need to work on merging this PR. Therefore, please re-open this PR.

**Comment by nsoranzo** [comment]
I supposed you wanted to open a clean PR after the others are merged, but I can surely reopen this, but I'll mark it WIP in the mean time.

**Comment by VJalili** [comment]
@nsoranzo Thanks.
